src/iop/highlights.c

```diff
@@ -199,21 +200,56 @@ int process_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, cl_m
                const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
 {
   dt_iop_highlights_data_t *d = (dt_iop_highlights_data_t *)piece->data;
+  dt_iop_highlights_gui_data_t *g = (dt_iop_highlights_gui_data_t *)self->gui_data;
   dt_iop_highlights_global_data_t *gd = (dt_iop_highlights_global_data_t *)self->global_data;
 
-  cl_int err = -999;
-  cl_mem dev_xtrans = NULL;
-
+  const uint32_t filters = piece->pipe->dsc.filters;
   const int devid = piece->pipe->devid;
   const int width = roi_in->width;
   const int height = roi_in->height;
 
+  const gboolean fullpipe = (piece->pipe->type & DT_DEV_PIXELPIPE_FULL) == DT_DEV_PIXELPIPE_FULL;
+  const gboolean visualizing = (g != NULL) ? g->show_visualize && fullpipe : FALSE;
+
+  cl_int err = DT_OPENCL_DEFAULT_ERROR;
+  cl_mem dev_xtrans = NULL;
+
+  // this works for bayer and X-Trans sensors
+  if(visualizing)
+  {
+    const gboolean unset = piece->pipe->dsc.temperature.coeffs[0] <= 0.0f;
```

Comment:
I would certainly clip each channel independently at least to be equivalent with the OpenCL implementation.

I'm thinking of something like this (not tested):

```
    const float *c = piece->pipe->dsc.temperature.coeffs;
    float clips[4] = { d->clip * (c[RED]   <= 0 ? 1.0f : c[RED]),
                       d->clip * (c[GREEN] <= 0 ? 1.0f : c[GREEN]),
...
```

Reply:
Yeah, thats maybe the better alternative. No problem for me (although i don't see a chance for any channel to be <= 0 and the others are not)

Reply:
done